_posts/2019-03-30-numba-stencil.md

```diff
@@ -0,0 +1,323 @@
+```
+
+
+Numba JIT
+---------
+
+**Docs:** http://numba.pydata.org/
+
+When we run this function on a NumPy array, we find that it is slow, operating
+at Python speeds.
+
+```python
+x = np.ones((100, 100))
+timeit f(x)
+527 ms ± 44.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
+```
+
+But if JIT compile this function with Numba, then it runs more quickly.
+
+```python
+@numba.njit
+def smooth(x):
+    return _smooth(x)
+
+%timeit smooth(x)
+70.8 µs ± 6.38 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
+```
+
+For those counting, that's over 1000x faster.
+
+Dask Array
+----------
+
+**Docs:** https://docs.dask.org/en/latest/array.html
+
+In these applications people often have many such arrays and they want to apply
+this function over all of them.  In principle they could do this with a for
+loop.
+
+```python
+from glob import glob
+import skimage.io
+
+for fn in glob('/path/to/*.png'):
+    img = skimage.io.imread(fn)
+    out = smooth(img)
+    skimage.io.imsave(fn.replace('.png', '.out.png'), out)
+```
+
+If they wanted to then do this in parallel they would maybe use the
+multiprocessing or concurrent.futures modules.  If they wanted to do this
+across a cluster then they could rewrite their code with PySpark or some other
+system.
+
+Or, they could use Dask array, which will handle both the pipelining and the
+parallelism (single machine or on a cluster) all while still looking mostly
+like a NumPy array.
+
+```python
+import dask_image
+x = dask_image.imread('/path/to/*.png')  # a large lazy array of all of our images
+y = x.map_blocks(smooth, dtype='int8')
+TODO: write
+```
+
+And then because each of the chunks of a Dask array are just NumPy arrays, we
+can use the `map_blocks` function to apply this function across all of our
+images, and then save them out.
+
+This is fine, but lets go a bit further, and discuss generalized universal
+functions from NumPy.
+
+
+Generalized Universal Functions
+-------------------------------
+
+**Numba Docs:** https://numba.pydata.org/numba-doc/dev/user/vectorize.html
+
+**NumPy Docs:** https://docs.scipy.org/doc/numpy-1.16.0/reference/c-api.generalized-ufuncs.html
+
+A generalized universal function (gufunc) is a normal function that has been
+annotated with typing and dimension information.  For example we can redefine
+our `smooth` function as a gufunc as follows:
+
+```python
+@numba.guvectorize(
+    [(numba.int8[:, :], numba.int8[:, :])],
+    '(n, m) -> (n, m)'
+)
+def smooth(x, out):
+    out[:] = _smooth(x)
+```
+
+This function knows that it consumes a 2d array of int8's and produces a 2d
+array of int8's of the same dimensions.
+
+This sort of annotation is a small change, but it lets other systems like Dask
+enough information to use it intelligently.  Rather than call functions like
+map_blocks, we can just use the function directly, as if our Dask Array was
+just a very large NumPy array.
+
+```python
+# Before gufuncs
+y = x.map_blocks(smooth, dtype='int8')
```

Comment:
Might be worth mentioning the `mean_filter` function in `dask-image`.

Reply:
Will do

Reply:
Actually, I'm not finding that function in dask-image

Reply:
I think it's `dask_image.ndfilters.uniform_filter`, with `size=3`.

Reply:
Ah sorry appears I have misremembered the name. Thanks @mrocklin for catching that and @pentschev for providing the correct function.

Reply:
Thanks!  They're about the same performance.  I've added a note.